### ISSUE TITLE
Make loading parts of checkpointing more flexible

### ISSUE THREAD

**Issue by moinnadeem** [mosaicml/composer#283]
## 🚀 Feature Request
Our original checkpointing design was built with the motivation of _pausing training_, and then resuming training for the same job. 
Then, we needed to add the ability to restore the weights without restoring the state to provide a good initialization for a new run.

Now, we are encountering some new situations where we want to restore _part_ of state. To motivate this by an example: I have pretrained BERT using some surgery module, such as replacing the activation function. When I run fine-tuning, I need the same surgery applied, since the model's representations depend on it. So now we need to restore part of state, but the `trainer/global_step`, and other metrics are irrelevant.

To complicate matters even further, I'm not even sure I want _all_ algorithms applied during fine-tuning. For instance, I have a suspicion that I wouldn't want sequence length warmup, since the dataset is so small. Similarly, I wouldn't want to run a layer freezing schedule or stochastic depth during fine-tuning. I tempted to say that I want to run all surgery algorithms when loading a pretrained checkpoint, since I cannot think of counterexamples.

In short, this might need to be a bit more flexible than our previous changes to checkpoint loading. It could be that a hyperparameter called `load_surgery_algorithms` is enough, or we could provide a list of parts of State to recover. 

There is a broader question to identify use cases that might occur outside of just loading surgery algorithms to motivate recovering parts of state, as well as seeing if we can unify our API anywhere.

**Comment by hanlint** [comment]
Can we push this complexity to the developer for now e.g. in your entrypoint script, you can do:
```
trainer = Trainer(load_path='....')
self.state.algorithms = [a for a in self.state.algorithms if isinstance(a, ...)
trainer.fit()
```

**Comment by moinnadeem** [comment]
Ah, this works! It's not a blocker, since I've been working around it, but your solution is better.
